--- a/routers/web/web.go
+++ b/routers/web/web.go
@@ -273,7 +273,7 @@
 					ctx.Data["Title"] = ctx.Tr("auth.must_change_password")
 					ctx.Data["ChangePasscodeLink"] = setting.AppSubURL + "/user/change_password"
 					if ctx.Req.URL.Path != "/user/events" {
-						middleware.SetRedirectToCookie(ctx.Resp, setting.AppSubURL+ctx.Req.URL.RequestURI())
+						if ctx.Req.Header.Get("X-Forgejo-UI") != "1" && !strings.Contains(ctx.Req.Header.Get("Accept"), "application/json") { middleware.SetRedirectToCookie(ctx.Resp, setting.AppSubURL+ctx.Req.URL.RequestURI()) }
 					}
 					ctx.Redirect(setting.AppSubURL + "/user/settings/change_password")
 					return
@@ -317,7 +317,7 @@
 		if options.SignInRequired != nil && options.SignInRequired() {
 			if !ctx.IsSigned {
 				if ctx.Req.URL.Path != "/user/events" {
-					middleware.SetRedirectToCookie(ctx.Resp, setting.AppSubURL+ctx.Req.URL.RequestURI())
+					if ctx.Req.Header.Get("X-Forgejo-UI") != "1" && !strings.Contains(ctx.Req.Header.Get("Accept"), "application/json") { middleware.SetRedirectToCookie(ctx.Resp, setting.AppSubURL+ctx.Req.URL.RequestURI()) }
 				}
 				ctx.Redirect(setting.AppSubURL + "/user/login")
 				return
@@ -332,7 +332,7 @@
 		if !options.SignOutRequired && !ctx.IsSigned && ctx.InteractiveReauthenticationPossible &&
 			ctx.GetSiteCookie(setting.CookieRememberName) != "" {
 			if ctx.Req.URL.Path != "/user/events" {
-				middleware.SetRedirectToCookie(ctx.Resp, setting.AppSubURL+ctx.Req.URL.RequestURI())
+				if ctx.Req.Header.Get("X-Forgejo-UI") != "1" && !strings.Contains(ctx.Req.Header.Get("Accept"), "application/json") { middleware.SetRedirectToCookie(ctx.Resp, setting.AppSubURL+ctx.Req.URL.RequestURI()) }
 			}
 			ctx.Redirect(setting.AppSubURL + "/user/login")
 			return
